### ISSUE TITLE
SwiftProtobuf does not generate .pb.swift for pure extension protos, causing build to fail

### ISSUE THREAD

**Issue by steeve** [bazelbuild/rules_swift#26]
SwiftProtobuf does not generate any file when passed with a proto file that defines only extensions. While this is well and good, the issue rises when the proto rule declares it, and thus because the file is not generated by the protoc plugin, fails because the `.pb.swift` is not created.

One such file is https://github.com/gogo/protobuf/blob/master/gogoproto/gogo.proto.

Perhaps the simplest way would be to wrap the plugin and make it generate a file if none was created?

**Comment by allevato** [comment]
The Swift protobuf generator doesn't have any logic that would suppress generation if a file only contained extensions; we generate a .pb.swift for every .proto file, and I just verified this locally. Based on the conversation in https://github.com/bazelbuild/rules_swift/pull/25#discussion_r196234240, I believe the real issue here is that we're not generating files properly for cross-repository references.

**Comment by steeve** [comment]
Perhaps you may want to hold on #25 the since it may hide another issue?

**Comment by allevato** [comment]
The fix in #25 is still valuable on its own since it fixes the issue for the WKTs that we filter, and cross-repo references appear to not have been working anyway before that fix (I just checked). So even if there's some relationship, I'm fine with having that fix go in separately. (And I already merged it 🙂 )

**Comment by allevato** [comment]
So this is interesting:

```
SUBCOMMAND: # @com_google_protobuf//:descriptor_proto [action 'Generating Swift sources for @com_google_protobuf//:descriptor_proto']
(cd /private/var/tmp/_bazel_allevato/f44dcef1b55f1cbd869234549890f1c2/execroot/build_bazel_rules_swift && \
  exec env - \
  bazel-out/host/bin/external/build_bazel_rules_swift/tools/mkdir_and_run/mkdir_and_run bazel-out/darwin-fastbuild/bin/e bazel-out/host/bin/external/com_google_protobuf/protoc '--plugin=protoc-gen-swift=bazel-out/host/bin/external/com_github_apple_swift_swift_protobuf/ProtoCompilerPlugin' '--swift_out=bazel-out/darwin-fastbuild/bin/e' '--swift_opt=FileNaming=FullPath' '--swift_opt=Visibility=Public' '--swift_opt=ProtoPathModuleMappings=bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/descriptor_proto.protoc_gen_swift_modules.asciipb' --descriptor_set_in bazel-out/darwin-fastbuild/genfiles/external/com_google_protobuf/descriptor_proto-descriptor-set.proto.bin google/protobuf/descriptor.proto)
```

Specifically, this part:

bazel-out/host/bin/external/build_bazel_rules_swift/tools/mkdir_and_run/mkdir_and_run **bazel-out/darwin-fastbuild/bin/e** bazel-out/host/bin/external/com_google_protobuf/protoc

It looks like the path string manipulation done in https://github.com/bazelbuild/rules_swift/blob/master/swift/internal/swift_protoc_gen_aspect.bzl#L106 doesn't produce the right result for cross-repo references. Digging further, the path to the .pb.swift file isn't even being generated correctly.

For example.pb.swift, we have:

`bazel-out/darwin-fastbuild/bin/examples/xplatform/proto/example_proto.protoc_gen_swift/examples/xplatform/proto/example.pb.swift`

For descriptor.pb.swift, we have:

`bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/com_google_protobuf/google/protobuf/descriptor.pb.swift`

In other words, there is no `.protoc_gen_swift` folder for the external references, and that screws up the string processing later that tries to extract that directory.

**Comment by allevato** [comment]
Yup, when we take an external file's `short_path` inside `_pbswift_file_path`, we end up with a path of the form `../repo_name/path/to/file`, and that's screwing it up.

I think what we want here is: `bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/descriptor_proto.protoc_gen_swift/google/protobuf/descriptor.pb.swift`.

**Comment by steeve** [comment]
You could try to use the workspace relative name also?

**Comment by allevato** [comment]
Yes, turns out that's the right fix here. I was initially concerned that it wouldn't be enough to handle the external repo output directory, but since we're in an aspect, all actions have their files automatically declared in the right place in the output tree. 👍 

PR coming up.
